bentoml/yatai/deployment/azure_function/operator.py

```diff
@@ -0,0 +1,724 @@
+    YataiDeploymentException,
+)
+from bentoml.yatai.proto.deployment_pb2 import (
+    ApplyDeploymentResponse,
+    DeploymentState,
+    DescribeDeploymentResponse,
+    DeleteDeploymentResponse,
+)
+from bentoml.yatai.proto.repository_pb2 import GetBentoRequest
+from bentoml.yatai.status import Status
+from bentoml.configuration import LAST_PYPI_RELEASE_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+def ensure_azure_cli_available_and_login_or_raise():
+    try:
+        account_info = _call_az_cli(
+            command=['az', 'account', 'show'], message='show Azure account'
+        )
+        if not account_info['user'] or not account_info['homeTenantId']:
+            # https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest
+            # Attempt to login with azure service principal values from bentoml
+            # configuration.
+            service_principal_app_id = config.get('yatai_service').get(
+                'azure_service_principal_app_url', None
+            )
+            service_principal_cert = config('yatai_service').get(
+                'azure_service_principal_cert', None
+            )
+            service_principal_tenant = config('yatai_service').get(
+                'azure_service_principal_tenant', None
+            )
+            if (
+                service_principal_app_id
+                and service_principal_cert
+                and service_principal_tenant
+            ):
+                service_principal_login_result = _call_az_cli(
+                    command=[
+                        'az',
+                        'login',
+                        '--service-principal',
+                        '--user-name',
+                        service_principal_app_id,
+                        '--password',
+                        service_principal_cert,
+                        '--tenant',
+                        service_principal_tenant,
+                    ],
+                    message='login Azure CLI with service principal account',
+                )
+                if service_principal_login_result['state'] != 'Enabled':
+                    raise BentoMLException(
+                        'Azure service principal account is not enabled'
+                    )
+            else:
+                raise BentoMLException('Azure CLI is not logged in')
+    except FileNotFoundError:
+        raise MissingDependencyException(
+            'azure cli is required for this deployment. Please visit '
+            'https://docs.microsoft.com/en-us/cli/azure/install-azure-cli '
+            'for instructions'
+        )
+
+
+def _init_azure_function_project(
+    azure_function_project_dir, bento_path, bento_pb, azure_function_config
+):
+    try:
+        shutil.copytree(bento_path, azure_function_project_dir)
+        with open(os.path.join(azure_function_project_dir, 'host.json'), 'w') as f:
+            f.write(AZURE_FUNCTION_HOST_JSON)
```

Comment:
if it's not a template, put it to a separate file and simply use `shutil.copyfile`